### ISSUE TITLE
Retain Cycle caused by the retention of download token

### ISSUE THREAD

**Issue by SteinX** [SDWebImage/SDWebImage#2855]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios
 Platform Version        | Nothing to do with the version
 SDWebImage Version      | 5.0.1
 Integration Method      | cocoapods
 Xcode Version           | Xcode 11
 Repro rate              | all the time (100%)
 Repro with our demo prj | yes
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

The retention of the cancel token inside the **completion block** or **progress block** will cause the retain cycle.

```objc
    __block id downloadToken;
    SDInternalCompletionBlock completionHandler = ^(UIImage * _Nullable image, NSData * _Nullable data, NSError * _Nullable error, SDImageCacheType cacheType, BOOL finished, NSURL * _Nullable imageURL) {
        // do something with downloadToken
    };

    downloadToken = [self.manager loadImageWithURL:URL
                                                options:options
                                               progress:progressHandler
                                              completed:completionHandler];
```

![screen_shot](https://user-images.githubusercontent.com/3668417/65829778-775fe800-e2db-11e9-99d5-62fdf3f3ff72.jpg)

**Comment by dreampiggy** [comment]
This retain cycle is caused by yourself. SDWebImage itself has no retain cycle issue.

For best practice, you shoud always set `self.token = nil` when you finished downloading, or when you dismiss your viewController. You can also use the weak reference to break this.

**Comment by dreampiggy** [comment]
See more here:https://github.com/SDWebImage/SDWebImage/wiki/Common-Problems#handle-self-capture-in-completion-block

**Comment by SteinX** [comment]
@dreampiggy Well, I'd like to mention that the download token is **a local variable** which is **only captured by the completion block** passed to the SDWebImageManager.  And `self` has been captured nowhere in my case (I do not use `self` in any block)

BTW, even though by declaring local variable with `__block __weak id downloadToken` can fix this issue, normally, no one can imagine that the capture of the token inside the completion block will make any problem without investigating in source code.

**Comment by dreampiggy** [comment]
The `downloadToken` is implmenetation detail. Because we support [Custom Loader](https://github.com/SDWebImage/SDWebImage/wiki/Advanced-Usage#custom-loader-50), which means, this can even be a random `NSOperation` class. We can not, because we can not limit the implementation detail for custom loader, that the returned value **is not retained by anything else**.

As a user, you should folllow the same correct way for this, use the weak reference, or assign it to nil after usage.

**Comment by dreampiggy** [comment]
But, for this case you mentioned about our own class, there are some implementation limit that we can not using weak. Let me explain the design for you if you don't understand the reason.

We used to have this feature: When you request same URL from different calls, all of them using the same `URLSessionTask`. To identify each original completion to callback, we retain the progressBlock and completionBlock in the `SDWebImageDownloaderOperation` and produce a dictionary. And the `SDWebImageDownloadToken` keep a strong reference to that dictionrary as `id downloadOperationCancelToken`.

There are already some code comments about this: See this line: https://github.com/SDWebImage/SDWebImage/blob/master/SDWebImage/Core/SDWebImageManager.m#L241

**Comment by dreampiggy** [comment]
This can be a consideration in 5.2.x version. For example, we can make some refactory to avoid this, including:

1. Find another way for generating the `downloadOperationCancelToken`, which does not retain the `progressBlock` or `completionBlock` but can aslo be identifiable which progressBlock is associated for this token (sounds impossible ?)
2. Try  to use the weak reference from `SDWebImageDownloadToken -> downloadOperationCancelToken`, this may be a simple solution, but I'm afriad this will break some exisiting logic for lifecycle.

**Comment by SteinX** [comment]
Thanks a lot for your patience and explanation.

I actually understand what's happening under the hood, however, from a ordinary user side, it just hard to aware of the potential memory leak by just capturing in the download token in the completion block without having the knowledge of implementation detail. That's why I raise this issue. 
Personally, it is not util I started debugging the memory issue that I noticed this thing happened.

Appreciate to your quick action to get it fixed.
